homeassistant/components/qingping/config_flow.py

```diff
@@ -0,0 +1,117 @@
+
+from .const import DOMAIN
+
+# How long to wait for additional advertisement packets if we don't have the right ones
+ADDITIONAL_DISCOVERY_TIMEOUT = 60
+
+
+class QingpingConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for qingping."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._discovery_info: BluetoothServiceInfoBleak | None = None
+        self._discovered_device: DeviceData | None = None
+        self._discovered_devices: dict[str, str] = {}
+
+    async def _async_wait_for_full_advertisement(
+        self, discovery_info: BluetoothServiceInfoBleak, device: DeviceData
+    ) -> BluetoothServiceInfoBleak:
+        """Sometimes first advertisement we receive is blank or incomplete. Wait until we get a useful one."""
+        if device.supported(discovery_info):
+            return discovery_info
+        return await async_process_advertisements(
+            self.hass,
+            device.supported,
+            {"address": discovery_info.address},
+            BluetoothScanningMode.ACTIVE,
+            ADDITIONAL_DISCOVERY_TIMEOUT,
+        )
+
+    async def async_step_bluetooth(
+        self, discovery_info: BluetoothServiceInfoBleak
+    ) -> FlowResult:
+        """Handle the bluetooth discovery step."""
+        await self.async_set_unique_id(discovery_info.address)
+        self._abort_if_unique_id_configured()
+        device = DeviceData()
+        try:
+            self._discovery_info = await self._async_wait_for_full_advertisement(
+                discovery_info, device
+            )
+        except asyncio.TimeoutError:
+            return self.async_abort(reason="not_supported")
+        self._discovery_info = discovery_info
+        self._discovered_device = device
+        return await self.async_step_bluetooth_confirm()
+
+    async def async_step_bluetooth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Confirm discovery."""
+        assert self._discovered_device is not None
+        device = self._discovered_device
+        assert self._discovery_info is not None
+        discovery_info = self._discovery_info
+        title = device.title or device.get_device_name() or discovery_info.name
+        if user_input is not None:
+            return self.async_create_entry(title=title, data={})
+
+        self._set_confirm_only()
+        placeholders = {"name": title}
+        self.context["title_placeholders"] = placeholders
+        return self.async_show_form(
+            step_id="bluetooth_confirm", description_placeholders=placeholders
+        )
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the user step to pick discovered device."""
+        if user_input is not None:
+            address = user_input[CONF_ADDRESS]
+            await self.async_set_unique_id(address, raise_on_progress=False)
```

Comment:
Looks like we need this to guard against multiple concurrent user flows creating an entry with the same address.
```suggestion
            await self.async_set_unique_id(address, raise_on_progress=False)
            self._abort_if_unique_id_configured()
```

Reply:
Good catch. We probably need to fix that in a few more places.  Will audit for that

Reply:
Oh its unreachable because we don't offer the device if its already configured in the `vol.In`

But still good to add because they could delete it in another window between steps

Reply:
https://github.com/home-assistant/core/pull/76624

Reply:
0f4b7fd330e7a573e5cda3738b98d981d017d9af